k2/csrc/hash.h

```diff
@@ -0,0 +1,281 @@
+unsigned long long int __forceinline__ __host__ __device__ AtomicCAS(
+    unsigned long long int* address,
+    unsigned long long int compare,
+    unsigned long long int val) {
+#ifdef __CUDA_ARCH__
+  return atomicCAS(address, compare, val);
+#else
+  *address = val;
+  return compare;
+#endif
+}
+
+/*
+  How Hash32 works:
+    - It's a map from uint32_t to int32_t.  (All keys except
+      ~0 == -1 == UINT32_MAX are allowed).
+    - Each bucket has a struct { uint32_t key; int32_t value; }
+    - If there is no value present, we set the key to ~0 == -1.
+    - The number of buckets is a power of 2 provided by the user.
+    - When accessing hash[key], we use bucket_index == key % num_buckets,
+      leftover_index = 1 | ((key * 2) / num_buckets).  This is
+      leftover part of the index times 2, plus 1.
+    - If the bucket at `bucket_index` is occupied, we look in locations
+      `bucket_index + n * leftover_index` for n = 1, 2, ....;  this choice
+      ensures that if multiple keys hash to the same bucket, they don't
+      all access the same sequence of locations; and leftover_index being
+      odd ensures we eventually try all locations (of course for reasonable
+      hash occupancy levels, we shouldn't ever have to try more than two
+      or three).
+    - When deleting values from the hash you must delete them all at
+      once (necessary because there is no concept of a "tombstone".
+
+  You use it by: constructing it, obtaining its Accessor with GetAccessor(), and
+  inside kernels (or host code), calling functions Insert(), Find() or Delete()
+  of the Accessor object.  There is no resizing; sizing it correctly is the
+  caller's responsibility and if the hash gets full the code will just loop
+  forever (of course it will get extremely slow before it reaches that point).
+*/
+class Hash32 {
+ public:
+
+  Hash32(ContextPtr c, int32_t num_buckets): data_(c, num_buckets, ~(uint64_t)0),
+                                             buckets_num_bitsm1_(0) {
+    K2_CHECK_GT(num_buckets, 64);
+    int32_t n = 2;
+    for (; n < num_buckets; n *= 2, buckets_num_bitsm1_++);
+    K2_CHECK_EQ(num_buckets, 2 << buckets_num_bitsm1_)
+        << " num_buckets must be a power of 2.";
+  }
+
+  // Shallow copy
+  Hash32 &operator=(const Hash32 &src) = default;
+  // Copy constructor (shallow copy)
+  Hash32(const Hash32 &src) = default;
+
+  ContextPtr &Context() { return data_.Context(); }
+
+  union Element {
+    uint64_t i;
+    struct {
+      uint32_t key;   // note: if ~key == 0, i.e. key == UINT32_MAX, it means
+      // there is no entry here.
+      int32_t value;  // value stored.  can be any int32_t.
+    } p;
+  };
+
+  class Accessor {
+   public:
+    Accessor(uint64_t *data,
+             uint32_t num_buckets_mask,
+             int32_t bucket_num_bitsm1) :
+        data_(data), num_buckets_mask_(num_buckets_mask),
+        bucket_num_bitsm1_(bucket_num_bitsm1) { }
```

Comment:
Do we need to check that

```cpp
K2_CHECK_EQ(num_buckets_mask_ + 1,  1 << (bucket_num_bitsm1 + 1));
```

Reply:
I checked this in the constructor of Hash32.